Resolve #112 Build Under Windows

mdouaihy opened the pull request:
## Which problem is this PR solving?
- Resolves  #112 

## Short description of the changes
* Move to OpenTracing 1.5.0 that builds under Windows
* Change link of OpenTracing to dynamic. With Static linking of OpenTracing, we end up with multi definition of OT Symbols. This is not acceptable especially for the OT Global Tracer
* Fix OS specificities
** Networking API
** Windows headers
* Fix some tests with random behavior that failed on Windows
* Fix some MSVC specific compilation
* Add appveyor build file.

Signed-off-by: Mehrez Douaihy <mehrez.douaihy@gmail.com>

rnburn commented on CMakeLists.txt:
```diff
@@ -51,11 +60,12 @@ else()
 # OpenTracingConfig.cmake file
 find_package(OpenTracing CONFIG REQUIRED)
-list(APPEND LIBS OpenTracing::opentracing-static)
+list(APPEND LIBS OpenTracing::opentracing)
```
Could OpenTracing static linking be provided as an option? It's necessary to make the plugin work with projects that are themselves statically linked or in environments without the c++ standard library installed.

See https://stackoverflow.com/q/47841812/4447365 and https://github.com/envoyproxy/envoy/pull/2252

mdouaihy replied:
The problem I faced with the static link to Open Tracing was that the OT Global Tracer was defined many times. For example, it was defined in the unit test executable and in the Jaeger library.

I can change the code to keep the link as before on Linux and the dynamic link on Windows.

What do you think?

rnburn replied:
I'd be fine with that.

I think there are scenarios where you'd want to link dynamically on linux also, but up to @isaachier how he'd like to support them.

isaachier replied:
Hey I must have totally missed this PR. Will look ASAP.

isaachier replied:
Ya ideally use `MSVC` or the like in `CMakeLists.txt` to make this decision.

mdouaihy replied:
changed. With MSVC, a shared library is used. Otherwise, it'll be a static one.

isaachier commented on CMakeLists.txt:
```diff
@@ -208,13 +219,35 @@ cmake_dependent_option(BUILD_SHARED_LIBS "Build shared libraries" ON
   list(APPEND JAEGERTRACING_LIBS jaegertracing)
+
+  if (MSVC)
+    # Copy the shared opentracing shared library next to jaeger tracing shared library
```
I understand the reason for this, but IDK if this is the ideal solution. Please ask @ruslo in the Hunter gitter chat.

ruslo replied:
I don't think it's related to Hunter, what is the exact reason of copying libraries?

isaachier replied:
Well this project is on Hunter at the moment. I think there is some issue with Windows, but awaiting more detail from @mdouaihy.

mdouaihy replied:
The problem I faced here is that the OT dynamic libraries were not accessible when running the Jaeger tests. On Linux, this works because of (or thanks to) the rpath.

ruslo replied:
> dynamic libraries were not accessible when running the Jaeger tests

You can modify PATH environment variable instead of copying libraries, something like that:
* https://github.com/cgold-examples/fruits/blob/081f76a970d3ebbc8d316ad8ef52994b8b4cad56/test/fruits/skin_off/CMakeLists.txt#L8-L27

mdouaihy replied:
ok, I will do that.

mdouaihy replied:
Changed. The path env variable is not set for the tests.

isaachier commented on CMakeLists.txt:
```diff
@@ -244,17 +284,21 @@ configure_file(
 if(JAEGERTRACING_BUILD_EXAMPLES)
   add_executable(app examples/App.cpp)
-  target_link_libraries(app PUBLIC ${JAEGERTRACING_LIB})
+  target_link_libraries(app PUBLIC jaegertracing)
```
Please undo this change.

mdouaihy replied:
change undone.

isaachier commented on CMakeLists.txt:
```diff
@@ -244,17 +284,21 @@ configure_file(
 
 if(BUILD_TESTING)
+
+  set(CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS ON)
```
Again, unsure if this is ideal. @ruslo should have some ideas here too.

ruslo replied:
It depends on your goal/approach. You can use `CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS` to export everything or use [GenerateExportHeader](https://cmake.org/cmake/help/latest/module/GenerateExportHeader.html) module and mark only particular classes/methods with `*_EXPORT`:
* https://github.com/cgold-examples/fruits/blob/081f76a970d3ebbc8d316ad8ef52994b8b4cad56/lib/fruits/rosaceae/Plum.hpp#L9
* https://github.com/cgold-examples/fruits/blob/081f76a970d3ebbc8d316ad8ef52994b8b4cad56/lib/fruits/rosaceae/Pear.hpp#L11

However I think if you decide to export everything then `set(CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS ON)` should be set once in top level `CMakeLists.txt` and there is no need to unset it, `set(CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS OFF)` can be removed.

isaachier replied:
So I think I had wanted to do the export header thing, just never got around to that. I was just wondering on your opinion about `set(CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS ON)`. It seems like something for `toolchain.cmake`. I know you do not believe `set(CMAKE_CXX_STANDARD 11)` is a good idea in CMakeLists.txt. I wonder if this could be a similar dangerous option.

mdouaihy replied:
I had a dffierent approach at the beginning where I exported the symbols one by one. I found that fastidious and ended with a huge list. That's why I adopted the approach of exporting all symbols.

isaachier replied:
I'm thinking of going the route that jaeger-client-java did. Essentially, why export any of the symbols other than perhaps the config? The OpenTracing API is the only public API we recommend to users.

mdouaihy replied:
I think that this can be a good idea. However, this means that won't be able to test the Jaeger code directly but we'll have to go through the OpenTracing API.

isaachier commented on CMakeLists.txt:
```diff
@@ -289,18 +333,20 @@ if(BUILD_TESTING)
       $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/src>)
     target_link_libraries(
-      DynamicallyLoadTracerTest OpenTracing::opentracing-static GTest::main)
+      DynamicallyLoadTracerTest OpenTracing::opentracing GTest::main)
```
Please revert for non-Windows. I remember specifically having issues with this on Mac and/or Linux.

mdouaihy replied:
change undone for Linux.

isaachier commented on cmake/Hunter/config.cmake:
```diff
@@ -0,0 +1,2 @@
+
+hunter_config(opentracing-cpp VERSION ${HUNTER_opentracing-cpp_VERSION} CMAKE_ARGS BUILD_SHARED_LIBS=ON)
```
Again, don't believe this is ideal. Hunter already has `HUNTER_BUILD_SHARED_LIBS` to control this. Please remove.

mdouaihy replied:
Hello @isaachier, HUNTER_BUILD_SHARED_LIBS impacts all the libraries imported by Hunter whereas the above line configure only OpenTracing as a shared Library. If I am not mistaken, this is what is suggested here https://github.com/ruslo/hunter/wiki/example.custom.config.id.

Why do you think is a bad idea?

@ruslo, what do you suggest? When I use HUNTER_BUILD_SHARED_LIBS, all libraries are considered as shared. The problem is that yaml-cpp does not build in such case.

ruslo replied:
> When I use HUNTER_BUILD_SHARED_LIBS, all libraries are considered as shared. The problem is that yaml-cpp does not build in such case

Homogeneous approach (all static or all shared) is correct and easy to check, so `HUNTER_BUILD_SHARED_LIBS` should be used in most cases. If there are some troubles with `yaml-cpp` package then this package should be fixed.

isaachier replied:
@ruslo I have one case where that is not ideal. This package is used as a plugin for https://github.com/envoyproxy/envoy. To do that, I need one big shared object with all dependencies packaged into it. I'll admit that it is pretty rare to use this pattern.

mdouaihy replied:
One problem with YAML is that under MSVC, there is a problem with a macro definition. I issued a PR for that https://github.com/hunter-packages/yaml-cpp/pull/4

isaachier replied:
Great. I am happy to see this was merged. We will need to prepare a new release of Hunter to land this review, but let's finish this part first.

mdouaihy replied:
I did a new fix in Yaml-cpp this week. Hunter updated https://github.com/ruslo/hunter/releases/tag/v0.23.202

isaachier commented on src/jaegertracing/Compilers.h:
```diff
@@ -0,0 +1,31 @@
+/*
+ * Copyright (c) 2017 Uber Technologies, Inc.
```
Wrong year.

mdouaihy replied:
Shall I update to 2019?

mdouaihy replied:
udated to 2019

isaachier commented on src/jaegertracing/Compilers.h:
```diff
@@ -0,0 +1,31 @@
+#pragma warning(push)
+#pragma warning(disable : 4251)
+
+// Define NOMINMAX to inhibit definition of Macros min(a,b) and max(a,b) in
```
Interesting fix.

isaachier commented on src/jaegertracing/DynamicLoad.cpp:
```diff
@@ -19,31 +19,36 @@
-int OpenTracingMakeTracerFactory(const char* opentracingVersion,
-                                 const void** errorCategory,
-                                 void** tracerFactory)
+int JaegerTracingMakeTracerFactoryFct(const char* opentracing_version,
```
Why?

mdouaihy replied:
I will revert this change.

isaachier replied:
Thank you. This has been fixed on master. Try this `git rebase master`.

mdouaihy replied:
change reverted.

isaachier commented on src/jaegertracing/DynamicLoad.h:
```diff
@@ -0,0 +1,26 @@
+/*
+ * Copyright (c) 2017 Uber Technologies, Inc.
```
Year issues again.

isaachier commented on src/jaegertracing/DynamicLoadTest.cpp:
```diff
@@ -60,12 +61,14 @@ disabled: false
     void* tracerFactory = nullptr;
-    const auto rcode = OpenTracingMakeTracerFactory(
-        "0.0.0" /*invalid version*/, &errorCategory, &tracerFactory);
+    const auto rcode = JaegerTracingMakeTracerFactoryFct(
```
Why?

isaachier commented on src/jaegertracing/Span.h:
```diff
@@ -21,6 +21,8 @@
 #include <memory>
 #include <mutex>
 
+#include "jaegertracing/Compilers.h"
```
`Config.h` should have included this everywhere already.

mdouaihy replied:
Why would I include Config.h in Span.h when I just need the compiler header fixing the MSVC compilation?

isaachier replied:
I mean if `Compilers.h` is included in `Config.h`, then one of the includes below will bring `Compilers.h` here too. Please try to keep local imports at the end and sorted alphabetically. Don't rely on include ordering when you don't have to please.

isaachier commented on src/jaegertracing/TracerTest.cpp:
```diff
@@ -250,6 +249,10 @@ TEST(Tracer, testTracer)
+    // There is a problem here. After replacing the global tracer, the spans produced
+    // hold the last references to the tracer. After they are published, the tracer is
+    // destroyed. This is not the right moment to close the tracer.
+    tracer->Close();
```
Why do you believe that is an issue? Who cares when the old tracer is closed?

mdouaihy replied:
The problem is that the tracer is being destroyed. During that phase; the tracer will close the resources that are still used to report the the spans asynchronously.

isaachier replied:
If you see a crash, please report it as a separate issue. I would like to research this a little more. The real fix should never involve an explicit close. For now, please remove these fixes and we can rebase this on top of another PR for this issue.

mdouaihy replied:
Ok I see. In fact, the point was not fix the crash but to avoid having the crash until it's fixed.

I will deactivate this test for the moment under windows and report it as an issue with the details.

isaachier replied:
Sure, but I want to try fixing that bug first. If it takes more than a day or two, we can disable the test to land this PR.

mdouaihy replied:
Issue reported #127

yurishkuro replied:
similarly, add link to the issue in the comments under TODO:

isaachier commented on src/jaegertracing/baggage/BaggageTest.cpp:
```diff
@@ -105,7 +157,19 @@ TEST(Baggage, testRemoteRestrictionManagerFunctionality)
     mockAgent->addBaggageRestriction("abc", Restriction(true, 1));
-    std::this_thread::sleep_for(std::chrono::milliseconds(200));
+
+    //Wait for the Update Restrictions to occurr
```
Typo on `occurr`.

mdouaihy replied:
Oh!!! Sorry for that. I will fix it.

isaachier commented on src/jaegertracing/baggage/BaggageTest.cpp:
```diff
@@ -105,7 +157,19 @@ TEST(Baggage, testRemoteRestrictionManagerFunctionality)
+
+    //Wait for the Update Restrictions to occurr
+    //This is done by listening to the number of update published via the Metrics
+    //Typically, On Windows with Debug mode, the update happens after 700ms
```
Please add spaces before start of line comment.

isaachier commented on src/jaegertracing/baggage/BaggageTest.cpp:
```diff
@@ -105,7 +157,19 @@ TEST(Baggage, testRemoteRestrictionManagerFunctionality)
-    std::this_thread::sleep_for(std::chrono::milliseconds(200));
+
+    //Wait for the Update Restrictions to occurr
+    //This is done by listening to the number of update published via the Metrics
```
I think this brings in another component that is not needed. We can use another mechanism to wait for updates.

mdouaihy replied:
Hello @isaachier, what do you suggest to listen for the updates. Why do you think that listening for the metrics is a bad idea?

isaachier replied:
It isn't a bad approach, but if we have `InMemoryReporter` designed for this use case. Why not use that? I agree this test has race conditions, but I'd rather fix them instead of using another component.

mdouaihy replied:
I dont think that the test has a race condition. it's just that some of the treatment is done asynchronously by another thread. We just need to wait to observe the result of that treatment. The reporter seemed to me as a way to observe such result.

isaachier commented on src/jaegertracing/net/Socket.h:
```diff
@@ -169,22 +195,50 @@ class Socket {
-    int handle() { return _handle; }
+    HandleType handle() { return _handle; }
+
+    class OSResource {
```
Please make this class `private`.

mdouaihy replied:
done

isaachier commented on src/jaegertracing/platform/Endian.h:
```diff
@@ -31,6 +31,31 @@
+#		define be64toh(x) (x)
+#	else
+#		error byte order not supported
+#   endif
```
Please format the code changes using `scripts/clang-format.sh`.

isaachier commented on src/jaegertracing/testutils/MockAgent.cpp:
```diff
@@ -31,10 +31,36 @@
+    static size_t write(int socketHandle, const char* msg, size_t size)
+    {
+#ifdef WIN32
+        return ::send(socketHandle, msg, size, 0);
```
`send` and `recv` work on Linux/Mac too. You can just use the Windows approach for all platforms and delete the original `read`/`write` calls.

mdouaihy replied:
done

isaachier commented on src/jaegertracing/ThriftMethods.cpp:
```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright (c) 2017 Uber Technologies, Inc.
```
Copyright year is wrong.

mdouaihy replied:
2019?

isaachier commented on src/jaegertracing/ThriftMethods.cpp:
```diff
@@ -0,0 +1,71 @@
+
+namespace jaegertracing {
+
+// trivial constructors of BaggageRestrictionManager thrift generated code are not generated by MSVC
```
What?! Why not? This is why I am not so keen on supporting Windows.

mdouaihy replied:
I didnt dig a lot. For example, you can see that the copy ctor of BaggageRestrictionManager_getBaggageRestrictions_args is declared in the header but isnt defined.

mdouaihy replied:
How can we proceed on this point? I think we can follow the bug in parallel, unless the thrift will dropped for protobuf.

isaachier replied:
There is no issue. You are correct. See if this works without the `#ifdef _MSVC`. It might not just be a Windows issue.

ruslo commented on CMakeLists.txt:
```diff
@@ -27,6 +28,14 @@ if(CMAKE_CXX_COMPILER_ID MATCHES "Clang")
   set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-unused-private-field")
 endif()
 
+if(CMAKE_CXX_COMPILER_ID MATCHES "MSVC")
```
You can use `if(MSVC)` here.

mdouaihy replied:
done

isaachier commented:
@mdouaihy do you plan on updating this or are you busy?

mdouaihy commented:
Hello @isaachier, sorry for the delay to answer but I have been sick for the past weeks.

I will take your comments into consideration and update the PR soon.

isaachier commented:
Sorry to hear you were sick @mdouaihy. I based some of the code changes in #120 on some changes in dynamic loading you have here, so I appreciate the PR. Whenever you have a chance to update this, I'll be glad to help review it.

mdouaihy commented:
@rnburn, @isaachier, @ruslo 
Sorry for the late PR update. I took me some time but I was able finally to allocate time for it.

isaachier commented on CMakeLists.txt:
```diff
@@ -203,12 +223,31 @@ function(add_lib_deps lib)
 
+function(update_path_for_test atest)
+  if(WIN32)
+  # If dependent library is a DLL we have to add location of DLL to PATH
```
Can you please indent this block?

mdouaihy replied:
done

isaachier commented on src/jaegertracing/Config.h:
```diff
@@ -17,6 +17,8 @@
 #define JAEGERTRACING_CONFIG_H
 
+#include "jaegertracing/Compilers.h"
+
```
Please remove newline here to keep with other Jaeger headers.

mdouaihy replied:
done

isaachier commented on src/jaegertracing/LogRecord.h:
```diff
@@ -17,6 +17,8 @@
 #ifndef JAEGERTRACING_LOGRECORD_H
 #define JAEGERTRACING_LOGRECORD_H
 
+#include "jaegertracing/Compilers.h"
```
~~As I've written before, this should be unnecessary. For instance, `Tag.h` should already include `Config.h`, which includes `Compilers.h`. Rule of thumb: you don't need to include a header already included in `Config.h`.~~

I'm sorry to say I got confused between this config header and one in the C client I have been working on (https://github.com/isaachier/jaeger-client-c/blob/master/src/jaegertracingc/config.h.in). My mistake. Keep this include wherever you need it.

isaachier commented on src/jaegertracing/Span.h:
```diff
@@ -21,6 +21,8 @@
 #include <memory>
 #include <mutex>
 
+//#include "jaegertracing/Compilers.h"
```
Please remove this line.

mdouaihy replied:
done

isaachier commented on src/jaegertracing/Reference.h:
```diff
@@ -22,6 +22,8 @@
 #include <opentracing/propagation.h>
 
+#include "jaegertracing/Compilers.h"
+
```
Newline issue.

isaachier commented on src/jaegertracing/Tracer.h:
```diff
@@ -83,29 +84,7 @@ class Tracer : public opentracing::Tracer,
          const std::shared_ptr<logging::Logger>& logger,
          metrics::StatsFactory& statsFactory,
-         int options)
-    {
```
Not a bad idea. I tried to inline a lot of code, but this is cleaner.

isaachier commented on src/jaegertracing/TracerFactory.cpp:
```diff
@@ -18,6 +18,13 @@
+
+//std::string node_data::empty_scalar;
+}
+}  // namespace YAML
```
Please remove this block entirely.

isaachier commented on src/jaegertracing/Transport.h:
```diff
@@ -17,6 +17,8 @@
 #ifndef JAEGERTRACING_TRANSPORT_H
 #define JAEGERTRACING_TRANSPORT_H
 
+#include "jaegertracing/Compilers.h"
```
Should appear last, where Jaeger headers always go. I believe `.clang-format` describes the ideal ordering.

isaachier commented on src/jaegertracing/UDPTransport.h:
```diff
@@ -17,6 +17,8 @@
 #define JAEGERTRACING_UDPTRANSPORT_H
 
+#include "jaegertracing/Compilers.h"
+
```
Please remove space.

isaachier commented on src/jaegertracing/UDPTransportTest.cpp:
```diff
@@ -100,7 +100,7 @@ TEST(UDPTransport, testExceptions)
     };
     for (auto type : exceptionTypes) {
-        MockUDPTransport sender(net::IPAddress(), 0, type);
+        MockUDPTransport sender(net::IPAddress::v4("localhost", 0), 0, type);
```
Why isn't this the default constructor?

mdouaihy replied:
because the behavior on windows is a bit different.
In `UDPClient::UDPClient`, `v_socket.connect(_serverAddr);` throws an exception: 
**Cannot connect socket to remote address { family=2, addr=0.0.0.0, port=0 }**

isaachier commented on src/jaegertracing/baggage/BaggageTest.cpp:
```diff
@@ -21,12 +21,61 @@
+    }
+
+    int64_t counterValue(const std::string& name,
+                         const metrics::StatsReporter::TagMap& tags)
```
`const` method?

isaachier commented on src/jaegertracing/baggage/BaggageTest.cpp:
```diff
@@ -105,7 +157,19 @@ TEST(Baggage, testRemoteRestrictionManagerFunctionality)
+    for (int sleepIndex = 0; sleepIndex < 10; ++sleepIndex) {
+        auto counterVal = testReporter.counterValue(
+            "jaeger.baggage-restrictions-update", { { "result", "ok" } });
+        std::this_thread::sleep_for(std::chrono::milliseconds(200));
```
Probably don't need to sleep if `counterValue` waits on mutex.

mdouaihy replied:
It's just a mutex and not a condition variable. We can have a spinning loop but we'll have to handle the timeout.

isaachier replied:
OK I understand the idea. If you'd like, you can add a condition variable too. Otherwise, this makes sense. Would prefer the condition variable to avoid sleeps, but will accept either solution.

isaachier commented on src/jaegertracing/baggage/BaggageTest.cpp:
```diff
@@ -21,12 +21,61 @@
+
+  private:
+    ValueMap _counters;
+    std::mutex _mutex;
```
Use `volatile` keyword to allow for usage in `const` methods.

mdouaihy replied:
mutable used.

isaachier commented on src/jaegertracing/baggage/RemoteRestrictionJSON.h:
```diff
@@ -17,6 +17,8 @@
 #define JAEGERTRACING_BAGGAGE_REMOTERESTRICTIONJSON_H
 
+#include "jaegertracing/Compilers.h"
+
```
Spacing...

isaachier commented on src/jaegertracing/net/IPAddress.cpp:
```diff
@@ -15,38 +15,92 @@
+struct CDeleter : public std::function<void(T*)> {
+    void operator()(T* ifAddr) const
     {
         if (ifAddr) {
```
I think both functions might work on a null argument. You can move `if` statement into `#else`.

mdouaihy replied:
Sorry @isaachier but I didnt get your point here.

isaachier replied:
Probably works:
```
void operator()(ifaddrs* ifAddr) const
{
#ifdef WIN32
    free(ifAddr);
#else
    ::freeifaddrs(ifAddr);
#endif
}
```

Definitely works:
```
#ifdef WIN32
    free(ifAddr);
#else
    if (ifAddr) {
        ::freeifaddrs(ifAddr);
    }
```
#endif
}

isaachier commented on src/jaegertracing/net/IPAddress.cpp:
```diff
@@ -15,38 +15,92 @@
-IPAddress IPAddress::localIP(std::function<bool(const ifaddrs*)> filter)
+#else
+
+#include <ifaddrs.h>
```
I'd rather not include this mid-file. Can you please use `#ifndef WIN32` to wrap this include after the system includes at the top?

isaachier commented on src/jaegertracing/net/IPAddress.cpp:
```diff
@@ -58,6 +112,18 @@ IPAddress IPAddress::localIP(std::function<bool(const ifaddrs*)> filter)
     return IPAddress();
 }
 
+}
```
Can you mark with `// anonymous namespace`?

isaachier commented on src/jaegertracing/net/IPAddress.h:
```diff
@@ -17,11 +17,7 @@
-#include <sys/socket.h>
-#include <sys/types.h>
-#include <unistd.h>
+#include "jaegertracing/Compilers.h"
```
Please move to end of includes.

isaachier commented on src/jaegertracing/net/Socket.h:
```diff
@@ -169,22 +195,53 @@ class Socket {
     int _family;
     int _type;
+
+    bool isHandleInvalid(HandleType h)
```
`static`?

isaachier commented on src/jaegertracing/net/http/ResponseTest.cpp:
```diff
@@ -24,14 +24,40 @@
+    static size_t read(int socketHandle, char* buffer, size_t size)
+    {
+#ifdef WIN32
+        int readBytes = ::recv(socketHandle, buffer, size, 0);
```
Please use `recv`/`send` for non-Windows systems too.

isaachier commented on src/jaegertracing/samplers/GuaranteedThroughputProbabilisticSampler.h:
```diff
@@ -17,6 +17,8 @@
 #define JAEGERTRACING_SAMPLERS_GUARANTEEDTHROUGHPUTPROBABILISTICSAMPLER_H
 
+#include "jaegertracing/Compilers.h"
+
```
Remove newline please.

isaachier commented on src/jaegertracing/samplers/SamplerTest.cpp:
```diff
@@ -39,7 +39,7 @@ constexpr auto kTestOperationName = "op";
 constexpr auto kTestDefaultSamplingProbability = 0.5;
 constexpr auto kTestMaxID =
-    static_cast<long double>(std::numeric_limits<uint64_t>::max()) / 2.0 + 1;
+    std::numeric_limits<uint64_t>::max() / 2 + 1;
```
@rnburn can explain the issue here (#6).

mdouaihy replied:
Unless I am missing something, in the issue #6, the problem occurs because we cast to double, which rounds up to a higher value. This latter, when casted back to  uint64_t cause the overflow.
Here, we are dividing an unsigned int by 2. I dont see why it can (over/under)flow.

isaachier replied:
Indeed you are correct. My mistake.

isaachier commented on src/jaegertracing/testutils/MockAgent.cpp:
```diff
@@ -31,10 +31,29 @@
 namespace testutils {
 namespace {
 
+struct SocketUtils {
```
What is this doing here again? And this one uses consistent behavior across platforms!

mdouaihy replied:
I will the remove those utility classes.
* I will replace write by ::send
* as for the read, I will keep two methods just to ajust the return value from signed to unsigned int.

isaachier replied:
OK great. Just check for errors (negative numbers) before casting please.

mdouaihy replied:
in fact, I used std::max(0, readBytesSize). This allowed t return 0 in case of error. Is that ok for you?

isaachier replied:
I don't understand. How does windows indicate a read was unsuccessful as opposed to successful but zero bytes read.

mdouaihy replied:
Here is the doc for both
* Linux
  * http://man7.org/linux/man-pages/man2/recvmsg.2.html
  * recv return an unsigned int
  * In case of error, it returns -1 (that will underflow to max int)
* Windows 
  * https://docs.microsoft.com/en-us/windows/desktop/api/winsock/nf-winsock-recv
  * recv return an int
  * In case of error, it returns -1

Now that I am answering you, I find myself thinking that I should test again the error value and return 0 or the value.

mdouaihy replied:
Here is the doc for both
* Linux
  * http://man7.org/linux/man-pages/man2/recvmsg.2.html
  * recv return an unsigned int
  * In case of error, it returns -1 (that will underflow to max int)
* Windows 
  * https://docs.microsoft.com/en-us/windows/desktop/api/winsock/nf-winsock-recv
  * recv return an int
  * In case of error, it returns -1

Now that I am answering you, I find myself thinking that I should test again the error value and return 0 or the value.

isaachier replied:
Looking through the original code, I see we handled the case of negative values correctly. My suggestion is just to replace `read`/`write` with `recv`/`send` where feasible. The return values will be handled correctly.

isaachier replied:
For the record, `ssize_t` is a *signed* size type, so there will be no underflow when it returns -1.

isaachier commented on src/jaegertracing/utils/UDPClientTest.cpp:
```diff
@@ -16,13 +16,13 @@
 
 #include "jaegertracing/net/IPAddress.h"
 #include "jaegertracing/net/Socket.h"
+
```
Please remove newline. I think I used the include-what-you-use tool on the project, which reordered all the headers. I'd rather have consistency with style going forward.

isaachier commented on src/jaegertracing/testutils/TUDPTransport.h:
```diff
@@ -62,23 +68,34 @@ class TUDPTransport
-                 reinterpret_cast<const ::sockaddr*>(&_clientAddr.addr()),
-                 _clientAddr.addrLen());
+#ifdef WIN32
+            reinterpret_cast<const char*>(buf),  // this cast is safe
```
That is fine with me. Any reason to make Windows specific and not on POSIX too?

mdouaihy replied:
On POSIX, send/sendto expect void*. It will be weird to cast to char*. What do you think?

isaachier replied:
Even if weird better than macros IMO.

mdouaihy replied:
I personally prefer the avoiding the cast when not needed.

isaachier commented on src/jaegertracing/utils/UDPClient.h:
```diff
@@ -59,7 +61,11 @@ class UDPClient : public agent::thrift::AgentIf {
             throw std::logic_error(oss.str());
         }
+#ifdef WIN32
+        const auto numWritten = ::send(_socket.handle(), reinterpret_cast<char*>(data), sizeof(uint8_t) * size, 0);
```
Same question regarding casting only on Windows.

mdouaihy replied:
changed to ::send for Linux and Windows.

isaachier commented:
This is all great. I understand you probably had to redo your changes entirely for this second review. From now on, if possible, please commit on top of the older changes and do not squash so we can see updates more easily.

FredrikAppelros commented:
Is work on this progressing? Do you need any help? Would love to see this merged ASAP.

mdouaihy commented:
Hi @FredrikAppelros, some tests are failing on Windows because if the bug described in issue #127.

I believe this is the last issue to solve in order to have this PR pushed.

isaachier commented:
> Hi @FredrikAppelros, some tests are failing on Windows because if the bug described in issue #127.
> 
> I believe this is the last issue to solve in order to have this PR pushed.

Hey I'll try to look this over tonight. If that is the only issue, I will allow a quick hack seeing as I have not come up with an easy fix for the bug.

solomonlu commented:
Hello everyone, can this PR continue? We are eager to use jaeger on Windows.

isaachier commented:
I'll try to see if this still merges and looks good.

isaachier commented:
I see I have a bunch of unresolved comments here. Please fix them and resubmit.

FredrikAppelros commented:
@isaachier: Are you referring to the style changes you requested or something more substantial when you mention unresolved comments?

As I offered before, I can help if I were given permission to push commits to this PR from @mdouaihy. The style changes should be easy enough, and I also have some fixes on top of this locally that I would like to see here.

One of the issues I had prevented me from getting this to compile since it could not pull in some dependency properly and the other makes sure that we can target the Windows 8.1 SDK.

I have not looked into the failing test case you guys mention though. Would be great if you could work out a solution for that.

isaachier commented:
@FredrikAppelros I have no control over the PR source, but you can always fork the windowsbuild branch from @mdouaihy and make the changes there. If you do that, we can open a PR based off of your version of windowsbuild and potentially land this without @mdouaihy's involvement. If @mdouaihy is free to help, then there is a better way to do this: make a PR onto @mdouaihy's windowsbuild branch and it will automatically update here when it lands there.

yurishkuro commented:
Or make smaller incremental PRs.

mdouaihy commented:
hello guys, I will recheck the changes requested by @isaachier. However, i still have the problem described in issue https://github.com/jaegertracing/jaeger-client-cpp/issues/127. That blocked me from going further.

isaachier commented:
@mdouaihy I was/am ready to accept this PR despite #127, especially because there is no easy fix there. You can use the local fix you had previously used. Just try to fix other comments and style suggestions.

efficks commented:
Hi, Is anybody have any news on this PR. Is someone needs help to make it merged? I offer my help if needed.

isaachier commented:
> Hi, Is anybody have any news on this PR. Is someone needs help to make it merged? I offer my help if needed.

Hey @efficks thanks for the offer. Last I heard, @mdouaihy was going to look over the changes. If you'd like to take over this PR, try pulling @mdouaihy's branch and writing a PR against @mdouaihy's fork. Then, if @mdouaihy accepts it on this branch, it will automatically update the PR.

mdouaihy commented:
Hello,
I am working again on the PR. However, due to update of msvc to 15.9, some dependencies dont compile anymore. I am following/working on them.

mdouaihy commented:
Hello @isaachier, I updated the PR with the changes you requested and rebased the version of the PR.

Would you please check and review it?

Sorry for the delay, but I had some personal stuff that prevented me from checking this PR before.

FredrikAppelros commented:
Any updates regarding this?

FredrikAppelros commented:
Sorry for nagging, just wondering if it is possible to get this reviewed?

mdouaihy commented:
Hello @isaachier, any news about the review?

yurishkuro commented on CMakeLists.txt:
```diff
@@ -210,12 +230,31 @@ function(add_lib_deps lib)
 
 if(BUILD_SHARED_LIBS)
   add_library(jaegertracing SHARED ${SRC})
+if (WIN32)
```
should this be indented more?

mdouaihy replied:
indeed.

yurishkuro commented on src/jaegertracing/Tracer.cpp:
```diff
@@ -214,4 +214,32 @@ Tracer::analyzeReferences(const std::vector<OpenTracingRef>& references) const
 }
 
+std::shared_ptr<opentracing::Tracer>
+Tracer::make(const std::string& serviceName,
```
Is this just moving this func from .h ?

mdouaihy replied:
yes.

yurishkuro commented on src/jaegertracing/UDPTransport.cpp:
```diff
@@ -27,6 +27,10 @@
 #include <thrift/transport/TBufferTransports.h>
 
+#ifdef _MSC_VER
+#pragma warning(disable : 4267)
```
it would be useful to add comments about why something is being disabled

mdouaihy replied:
ok

yurishkuro commented:
@isaachier hasn't been working on Jaeger for some time now, so he may not be around to review this PR. 

Is there a way to test that the Windows build is actually fixed? I see that `.appveyor.yml` is added, but there's no CI status posted from appveyor. Has the project been set up there?

mdouaihy commented:
Hi @yurishkuro, I configured a build for my fork at https://ci.appveyor.com/project/mdouaihy/jaeger-client-cpp/builds with appveyor configuration. I am thinking of keeping only Debug or Release instead of having them both (leading to doubled build time). What do you think?

yurishkuro commented:
@mdouaihy what do you mean by "keeping debug or release only"? keeping where?

we seem to have an official appvoyer account for this repo, but it has no builds since last year.

https://ci.appveyor.com/project/jaegertracing/jaeger-client-cpp/settings

Would you like to configure it to build for this PR? I sent you an invitation to that project.

mdouaihy commented:
Hi @yurishkuro, my question was about the types of builds to run under windows: Debug and/or Release. I ended up keeping only the Debug build since the release one does offer any added value.

As for the integration with Appveyor, I dont have permission on the github project. Can you check that the Github Appveyor app is well configured to run on PR?

yurishkuro commented:
I reinstalled AppVeyor GitHubApp.

I see this in the Events
![image](https://user-images.githubusercontent.com/3523016/64078809-77a19f00-ccad-11e9-8676-8cf1460d1adb.png)

But in the build history I only see older builds.

mdouaihy commented:
I think that this is because it's configured to run only on the master and not for the PR, no?

yurishkuro commented:
I don't think so.
![image](https://user-images.githubusercontent.com/3523016/64078860-3cec3680-ccae-11e9-9ea1-f99bfe92a36e.png)

![image](https://user-images.githubusercontent.com/3523016/64078867-4a092580-ccae-11e9-8fcc-e12e31b4d44f.png)

mdouaihy commented:
I pushed a small change to the PR. Appveyor was triggered.

yurishkuro commented on src/jaegertracing/TracerTest.cpp:
```diff
@@ -218,14 +219,15 @@ TEST(Tracer, testTracer)
             ->StartSpanWithOptions("test-span-with-default-system-timestamp",
                                    options)
             .release()));
+    /*
```
please add a comment, or better a link to an issue describing why this test is being disabled

mdouaihy replied:
ok. Issue inserted #167 and issue referenced in the code.

mdouaihy commented:
it seems that Appveyor does both: update the check and submit a message.

yurishkuro commented:
Hm, I don't get it, now the Travis status is gone. I was hoping to get both, testing on different OSes

mdouaihy commented:
hi @yurishkuro, maybe we need to reinstall travis ci?

mdouaihy commented:
hi @yurishkuro, any update?

yurishkuro commented:
I disabled and re-enabled the repo in Travis, and restarted the last build (which was for commit  d08082c, so it missed several latest commits).

yurishkuro commented:
Can you push another commit to this PR?

mdouaihy commented:
Thank you @yurishkuro !

yurishkuro commented:
thank YOU for your contribution.

FredrikAppelros commented:
Excellent! Will we see this in a release soon? :)

yurishkuro commented:
According to https://github.com/jaegertracing/jaeger-client-cpp/blob/master/RELEASE.md, releasing is just a tag?  If someone can create the pre-release PR (changelog changes, etc), then I can do a release.

mdouaihy commented:
PR #169 sent for that purpose.
